lib/Controller/GroupAvatarController.php

```diff
@@ -0,0 +1,176 @@
+	 */
+	public function postAvatar(): DataResponse {
+		$file = $this->request->getUploadedFile('file');
+
+		if (is_null($file)) {
+			return new DataResponse(
+				['message' => $this->l->t('No image file provided')],
+				Http::STATUS_BAD_REQUEST
+			);
+		}
+
+		if (
+			$file['error'] === 0 &&
+			is_uploaded_file($file['tmp_name']) &&
+			!Filesystem::isFileBlacklisted($file['tmp_name'])
+		) {
+			if ($file['size'] > 20 * 1024 * 1024) {
+				return new DataResponse(
+					['message' => $this->l->t('File is too big')],
+					Http::STATUS_BAD_REQUEST
+				);
+			}
+			$content = file_get_contents($file['tmp_name']);
+			unlink($file['tmp_name']);
+		} else {
+			return new DataResponse(
+				['message' => $this->l->t('Invalid file provided')],
+				Http::STATUS_BAD_REQUEST
+			);
+		}
+
+		try {
+			$image = new \OC_Image();
+			$image->loadFromData($content);
+			$image->readExif($content);
+			$image->fixOrientation();
+			if (!($image->height() === $image->width())) {
+				throw new NotSquareException($this->l->t('Avatar image is not square'));
+			}
+
+			if (!$image->valid()) {
+				return new DataResponse(
+					['data' => ['message' => $this->l->t('Invalid image')]],
+					Http::STATUS_BAD_REQUEST
+				);
+			}
+
+			$mimeType = $image->mimeType();
+			if ($mimeType !== 'image/jpeg' && $mimeType !== 'image/png') {
+				return new DataResponse(
+					['data' => ['message' => $this->l->t('Unknown filetype')]],
+					Http::STATUS_BAD_REQUEST
+				);
+			}
+
+			try {
+				$folder = $this->appData->getFolder('room-avatar');
+			} catch (NotFoundException $e) {
+				$folder = $this->appData->newFolder('room-avatar');
+			}
+			$token = $this->getRoom()->getToken();
+			$extension = explode('/', $mimeType)[1];
+			$folder->newFile($token . '.' . $extension, $image->data());
+			return new DataResponse();
+		} catch (NotSquareException $e) {
+			return new DataResponse(['message' => $e->getMessage()], Http::STATUS_BAD_REQUEST);
+		} catch (\Exception $e) {
+			$this->logger->error('Failed to post avatar', [
+				'exception' => $e,
+			]);
+
+			return new DataResponse(['message' => $this->l->t('An error occurred. Please contact your administrator.')], Http::STATUS_BAD_REQUEST);
+		}
+	}
+
+	/**
+	 * @PublicPage
+	 * @RequireParticipant
+	 * @RequireModeratorOrNoLobby
+	 */
+	public function getAvatar(): Response {
+		try {
+			$folder = $this->appData->getFolder('room-avatar');
+		} catch (NotFoundException $e) {
+			$folder = $this->appData->newFolder('room-avatar');
+		}
+		$token = $this->getRoom()->getToken();
+		if ($folder->fileExists($token . '.png')) {
+			$file = $folder->getFile($token . '.png');
+		} elseif ($folder->fileExists($token . '.jpeg')) {
+			$file = $folder->getFile($token . '.jpeg');
+		} else {
+			// @todo Implement fallback images
+			return new DataResponse([], Http::STATUS_NOT_FOUND);
+		}
+
+		$resp = new FileDisplayResponse($file);
+		$resp->addHeader('Content-Type', $file->getMimeType());
+		return $resp;
```

Comment:
Isn't one day a long time?

Reply:
User avatars are also cached 1 day. How often would you change a conversation avatar? Most of the chats I have basically set the image once and then never change it.

Reply:
I think that we will need to store the latest version of avatar in the room table.

When the participant A as a room moderator change the avatar
And the participant B retrieve the Avatar
And the participant A change the avatar a more time
The participant B need to receive the newest version of avatar.

If we set the cache without a query string defining the version of avatar, the participant B don't will see the newest avatar of room by a day.

@nickvergessen what you think to put the avatar version in room table? I suggest this table because avatar is an attribute of room.

Reply:
But it's the same for user avatars.
A day is totally fine. We can cachr it shorter if it is the fall back icon avatar, but i would not do effort to store a version of the avatar.
This is a very rare scenario and it is the same on basically every service.